### ISSUE TITLE
TSOA unable to ignore constructor parameters

### ISSUE THREAD

**Issue by WoH** [lukeautry/tsoa#394]
When documenting classes, tsoa documents both properties and all constructor parameters not marked private or protected. This leads to issues with classes like this:

```
export class TsoaError {
  /**
   * @ignore
   */
  readonly status: number

  constructor(status: number, public readonly message: string) { }
}
```

When TSOA parses this class, it correctly iterates model properties and filters the ignored property. Then it iterates the constructor parameters, which is perfect since I want to document the message. However, it re-adds the ignored status, since the filter for constructor parameters only ignores parameters explicitly marked `private` or `protected`.

See [here](https://github.com/lukeautry/tsoa/blob/master/src/metadataGeneration/typeResolver.ts#L578) and [here](https://github.com/lukeautry/tsoa/blob/master/src/metadataGeneration/typeResolver.ts#L658)

I would really like to change this to add only explicitly marked public parameters to the metadata. I'd argue he benefit of having the unmarked parameters added to the metadata is not worth the inability to remove them.

I'd happily open the PR :)

**Comment by dgreene1** [comment]
In TypeScript, the default accessor is public if you don’t explicitly define it’s accessor ((citation here)[https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers]). So if we were to make the change that you’re requesting, that would be hiding public members. This would be confusing and would likely result in other users filing bugs against tsoa.

So we’ll have to close this issue since it is behaving as expected. Unless you mark a member as private or protected (or use the ignore JSDoc (which I was actually surprised that tsoa had)) then the memeber will be both public in tsoa and the generated swagger.

**Comment by dgreene1** [comment]
But if you would like to submit a PR that fixes the fact that the ignore JSDoc tag isn’t being applied for constructor parameters, we’d accept that.

I just can’t accept this solution you described for the reasons I stated in the prior comment:
> “I would really like to change this to add only explicitly marked public parameters to the metadata.”

**Comment by WoH** [comment]
That is semi-correct. You can parse the modifiers and check.

```
  private hasPublicModifier(node: ts.Node) {
    return !node.modifiers || node.modifiers.every((modifier) => {
      return modifier.kind !== ts.SyntaxKind.ProtectedKeyword && modifier.kind !== ts.SyntaxKind.PrivateKeyword;
    });
  }
```

**Comment by WoH** [comment]
I should clarify, because I didn't explain the issue well I think.
> In TypeScript, the default accessor is public if you don’t explicitly define it’s accessor

I was referring to explicitly declare public in the constructor, not model properties.

While
```
export class TsoaError {
  constructor(public status: number) { }
}
```
adds the accessor (tsoaError.status), 
```
export class TsoaError {
  constructor(status: number) { }
}
```
does not.

**Comment by dgreene1** [comment]
Thank you. I understand the scenarios better. It's clear to me now that this is a bug and it has nothing to do with the ignore tag. Simply put, in this example, status should not be generated in the model:

```ts
export class TsoaError {
  constructor(status: number) { }
}
```

We would gladly accept bug-fix PR if there were a unit test that could pass for the following case:

This is the setup:
```ts

export class MixedAccessorModel {
  public explicitlyPublicClassMember: string;
  implicitlyPublicMember: string;
  private explicitlyPrivateMember: string;
  constructor(
      public explicitParamFromConstructor: number,
      tempConstructorParam: string,
      explicitlyPublicClassMember: string,
      implicitlyPublicMember: string,
      explicitlyPrivateMember: string,
  ) {
      this.explicitlyPublicClassMember = explicitlyPublicClassMember;
      this.implicitlyPublicMember = implicitlyPublicMember;
      this.explicitlyPrivateMember = explicitlyPrivateMember;
  }
}

const anError = new MixedAccessorModel(400, "temp", "myExplicitMemberValue", "myImplicitMemberValue", "thisIsPrivate");

// This is allowed
console.log(anError.explicitParamFromConstructor);
// This is allowed
console.log(anError.explicitlyPublicClassMember);
// This is allowed
console.log(anError.implicitlyPublicMember);
// This is not allowed
console.log(anError.tempConstructorParam); // Property 'tempConstructorParam' does not exist on type 'MixedAccessorModel'.
// This is not allowed
console.log(anError.explicitlyPrivateMember); // Property 'explicitlyPrivateMember' is private and only accessible within class 'MixedAccessorModel'.
```

And the unit test expectation in `definitions.spec.ts` would be something like:
```ts
const definition = getValidatedDefinition('MixedAccessorModel');

if (!definition.properties) { throw new Error(`Definition has no properties ${forSpec(currentSpec)}.`); }
if (!definition.properties.explicitParamFromConstructor) { throw new Error(`There was no 'explicitParamFromConstructor' property ${forSpec(currentSpec)}.`); }
expect(definition.properties.value.type).to.equal('string', forSpec(currentSpec));
if (!definition.properties.explicitlyPublicClassMember) { throw new Error(`There was no 'explicitlyPublicClassMember' property ${forSpec(currentSpec)}.`); }
expect(definition.properties.value.type).to.equal('string', forSpec(currentSpec));
if (!definition.properties.implicitlyPublicMember) { throw new Error(`There was no 'implicitlyPublicMember' property ${forSpec(currentSpec)}.`); }
expect(definition.properties).not.toHaveProperty('tempConstructorParam');
expect(definition.properties).not.toHaveProperty('explicitlyPrivateMember'); // note: I don't think toHaveProperty is a real chai assertion, so this is psuedo-code
```

**Comment by dgreene1** [comment]
P.s. I forgot to add a thank you @WoH for explain it. I don’t use classes much in TypeScript. I like doing a more data only / functional approach. Anyway, I think the case above should cover the bug and also maintains existing expectations.

**Comment by WoH** [comment]
@dgreene I found out about `@ignore` reading the source, maybe someone can document the behavior?

**Comment by dgreene1** [comment]
I approved the PR, so this is closed. We'll push that fix out to NPM soon. Btw, I'm not a big fan of the `@ignore` JSDoc annotation, and since we were able to solve the issue without even using `@ignore` then I think this is settled. I don't plan on documenting it's usage since I believe it is dangerous. If a property is public and it's ignored, then the property would come in as undefined and then the API would hit a runtime error whenever anyone used the ignore property. So if we were ever to document the ignore annotation, I would also like to see that come with an enhancement that only allows the ignore annotation to be used on an optional member. But we'd have to track that in a separate issue. But this issue is closed thanks to your excellent PR. :)
